compiler/test/stdlib/immutablearray.test.gr

```diff
@@ -0,0 +1,430 @@
+
+assert sum == 0
+
+let addToSum = n => {
+  sum += n
+}
+
+ImmutableArray.cycle(addToSum, 2, arr1)
+
+assert sum == 12
+
+let arr2 = fromList(["a", "b", "c"])
+let mut word = ""
+let addToWord = n => {
+  word = word ++ n
+}
+
+ImmutableArray.cycle(addToWord, 2, arr2)
+
+assert word == "abcabc"
+
+sum = 0
+
+ImmutableArray.cycle(sentinel, 0, arr1)
+
+assert sum == 0
+
+ImmutableArray.cycle(sentinel, -42, arr1)
+
+assert sum == 0
+
+// ImmutableArray.forEach
+
+let arr1 = fromList([1, 2, 3])
+let mut sum = 0
+
+let sentinel = _n => {
+  sum = 42
+}
+
+ImmutableArray.forEach(sentinel, fromList([]))
+
+assert sum == 0
+
+let addToSum = n => {
+  sum += n
+}
+
+ImmutableArray.forEach(addToSum, arr1)
+
+assert sum == 6
+
+let mut list = []
+ImmutableArray.forEach(x => list = [x, ...list], bltArr)
+assert List.reverse(list) == branchingLongTail
+
+// ImmutableArray.reverse
+
+assert ImmutableArray.reverse(fromList([])) == fromList([]) // Empty
+let testArr = fromList([1, 2, 3, 4])
+assert ImmutableArray.reverse(testArr) ==
+  fromList([4, 3, 2, 1]) // array is reversed properly
+assert testArr == fromList([1, 2, 3, 4]) // original array is not modified
+
+// ImmutableArray.toList
+
+assert ImmutableArray.toList(fromList([])) == []
+assert ImmutableArray.toList(fromList([1, 2, 3])) == [1, 2, 3]
+assert ImmutableArray.toList(fromList(branchingLongTail)) == branchingLongTail
+assert ImmutableArray.toList(fromList(noTail)) == noTail
+
+// ImmutableArray.map
+
+assert ImmutableArray.map(incr, fromList([])) == fromList([])
+assert ImmutableArray.map(incr, fromList([2, 3, 4])) == fromList([3, 4, 5])
+assert ImmutableArray.map(incr, bltArr) ==
+  fromList(List.map(incr, branchingLongTail))
+
+// ImmutableArray.contains
+
+let arr = fromList([1, 2, 3])
+
+assert ImmutableArray.contains(3, arr)
+assert !ImmutableArray.contains(4, arr)
+
+// ImmutableArray.find
+
+assert ImmutableArray.find(x => x == 2, arr) == Some(2)
+assert ImmutableArray.find(x => fail "Shoulnd't be called", fromList([])) ==
+  None
+assert ImmutableArray.find(x => false, arr) == None
+
+// ImmutableArray.findIndex
+
+assert ImmutableArray.findIndex(x => x == 1, arr) == Some(0)
+assert ImmutableArray.findIndex(x => x == 2, arr) == Some(1)
+assert ImmutableArray.findIndex(x => fail "Shouldn't be called", fromList(
+  []
+)) ==
+  None
+assert ImmutableArray.findIndex(x => false, arr) == None
+
+// ImmutableArray.product
+
+let arrA = fromList([1, 2])
+let arrB = fromList([2, 3, 4])
+let arrC = fromList([1])
+let arrD = fromList([2])
+
+assert ImmutableArray.product(arrA, arrB) ==
+  fromList([(1, 2), (1, 3), (1, 4), (2, 2), (2, 3), (2, 4)])
+assert ImmutableArray.product(arrC, arrB) == fromList([(1, 2), (1, 3), (1, 4)])
+assert ImmutableArray.product(arrC, arrD) == fromList([(1, 2)])
+assert ImmutableArray.product(fromList([]), fromList([])) == fromList([])
+
+// ImmutableArray.zip
+
+let arrA = fromList([1, 2, 3, 4, 5])
+let arrB = fromList([5, 4, 3, 2, 1])
```

Comment:
Are we supporting different length ImmutableArrays? We are changing the other `zip` functions for that in 0.6

Reply:
Yep, the behavior should be the same as `List.zip` where it has the length of the smaller array. I'll add a test for this